File: src/lean/fold-payload.ts
/**
 * `FoldCaptureFile` + `CapturedDay` → a `verified_cli day` request.
 *
 * Task #424. The wire format was decided by measurement first
 * (`lean/experiments/passfold-env-size.mts`): recorded answer tables for the
 * six mirror lookups, the day tables as data, and the road / walk solvers left
 * as shell callbacks so their 4.31 MiB/day of geometry never crosses. That is
 * 0.35 MiB/day steady state against the HSMM tenant's 33-40 (#411).
 *
 * # Floats cross as bit patterns, and so do the table KEYS
 *
 * Every float is a `Float64` bit pattern rendered as a decimal integer string,
 * the same convention as every other tenant, so both arms compare the same
 * doubles rather than two 6-decimal renderings of them.
 *
 * The lookup tables are keyed the same way, and that part is not cosmetic. A
 * recorded trace keys on `${lat}|${lon}|${radius}` — a JS number rendering.
 * Lean renders doubles differently, so a Lean-side key built from the same
 * double would not match the string, and a key that does not match is a MISS.
 * Re-keying on bits makes the two sides agree exactly or not at all.
 *
 * Recovering the double from the trace key is exact: JS `Number(String(x))`
 * round-trips, so `Number("51.502")` is bit-identical to the coordinate that
 * produced the key.
 *
 * # What a miss means
 *
 * `verified_cli day` aborts on an unanswered lookup rather than returning an
 * empty one, because empty is a real answer with a real meaning:
 * `LineMembership.scan` reads an empty served-station list as "line unknown"
 * and vetoes a journey that happened (#423). So a miss is a hard failure, and
 * it names the key. Run the CLI with `LEAN_ABORT_ON_PANIC=1` — measured, Lean's
 * `panic!` otherwise prints and CONTINUES with the default.
 *
 * A miss is a finding, not a gap in this encoder: it means the Lean fold asked
 * a question the TS cascade did not, which is a wiring divergence that output
 * comparison alone would not localise.
 *
 * THAT SENTENCE WAS FALSE FOR FOUR COMMITS, and the caveat is worth keeping.
 * It holds only because `answers` is now what the run itself answered. While
 * these tables were built from the fixture's `osmTrace`, a miss could equally
 * mean the encoder had handed Lean a narrower oracle than the TS arm used, and
 * nothing in the message distinguishes the two (#428). An encoder that decides
 * what the fold is allowed to ask has to be sure it is not the one failing.
 */

import type { CapturedDay } from "../cli/fixture-day.js";
import type { EnrichedSegment } from "../geo/enriched-segment.js";
import type { NominatimResult } from "../geo/osm.js";
import { DEFAULT_RADIUS_M } from "../geo/osm.js";
import type { OsmTrace } from "../geo/osm-adapter-recording.js";

/** `reverseGeocode`'s `zoom = 18` default, so a trace key that omitted it records
 *  the EFFECTIVE argument rather than a blank — the same choice `table3` makes
 *  for an omitted radius. */
const NOMINATIM_DEFAULT_ZOOM = 18;

import type { FoldCaptureFile } from "./fold-capture.js";

/** A `Float64` bit pattern as a decimal string — the bridge's float encoding. */
export function bits(x: number): string {
	const d = new DataView(new ArrayBuffer(8));
	d.setFloat64(0, x);
	return d.getBigUint64(0).toString();
}

const optBits = (x: number | null | undefined): string | null => (x === null || x === undefined ? null : bits(x));
const optStr = (x: string | null | undefined): string | null => (x === null || x === undefined ? null : x);

type Path = { lat: number; lon: number; ts: number }[] | undefined;
const path = (p: Path): string[][] | null =>
	p === undefined ? null : p.map((q) => [bits(q.lat), bits(q.lon), bits(q.ts)]);

/** One segment in the shape `Day.parseSeg` reads. */
export function encodeSeg(s: EnrichedSegment): unknown {
	return {
		startTs: s.startTs,
		endTs: s.endTs,
		mode: s.mode,
		refinedMode: optStr(s.refinedMode),
		confidence: bits(s.confidence),
		confidenceMargin: bits(s.confidenceMargin),
		avgSpeed: bits(s.avgSpeed),
		maxSpeed: bits(s.maxSpeed),
		linearity: bits(s.linearity),
		pointCount: s.pointCount,
		place: optStr(s.place),
		city: optStr(s.city),
		wayName: optStr(s.wayName),
		refinedReason: optStr(s.refinedReason),
		refinedKinds: s.refinedKinds ? [...s.refinedKinds] : [],
		centroidLat: optBits(s.centroidLat),
		centroidLon: optBits(s.centroidLon),
		// `focusPlaceId` is `string | number` in TS and `Option Int` in Lean.
		// Numeric ids are what the pipeline actually carries; a non-numeric one
		// would silently become NaN, so it is dropped rather than coerced.
		focusPlaceId: s.focusPlaceId === undefined || Number.isNaN(Number(s.focusPlaceId)) ? null : Number(s.focusPlaceId),
		needsReenrich: s.needsReenrich ?? false,
		vehicleKind: optStr(s.vehicleKind),
		roadCorridorFraction: optBits(s.roadCorridorFraction),
		displayTz: optStr(s.displayTz),
		snappedPath: path(s.snappedPath),
		matchedPath: path(s.matchedPath),
		walkMatchedPath: path(s.walkMatchedPath),
		walkSmoothedPath: path(s.walkSmoothedPath),
		biometrics: s.biometrics
			? {
					hrMean: optBits(s.biometrics.hrMean),
					hrMin: optBits(s.biometrics.hrMin),
					hrMax: optBits(s.biometrics.hrMax),
					hrStd: optBits(s.biometrics.hrStd),
					sampleCount: s.biometrics.sampleCount,
					overlapsSleep: s.biometrics.overlapsSleep,
					sleepFraction: bits(s.biometrics.sleepFraction),
					stepsTotal: optBits(s.biometrics.stepsTotal),
				}
			: null,
	};
}

/** Split an `osmTrace` key back into its numeric arguments. Exact: the key was
 *  built by JS number rendering, which round-trips. */
function keyNums(k: string): number[] {
	return k
		.split("|")
		.filter((s) => s !== "")
		.map(Number);
}

/** `[latBits, lonBits, radiusBits, answer]` from a trace section.
 *
 *  MEASURED, and it is why this takes a default: the trace keys on the radius
 *  the CALLER passed, and three of `velocity.ts`'s four `linesAtPoint` call
 *  sites pass none — so those entries are keyed `lat|lon|` while the adapter
 *  actually answered at `DEFAULT_RADIUS_M.linesAtPoint` (100 m). The Lean fold
 *  passes the radius explicitly, as a function of three arguments must, so it
 *  asks `lat|lon|100`: the same question in a different spelling, and without
 *  this it reads as a miss.
 *
 *  Filling in the default is recording the EFFECTIVE argument, the same choice
 *  `fold-capture.ts` makes for the `tzAt` fallback. It is not a widening: the
 *  omitted radius has exactly one value. */
function table3<T>(section: Record<string, T> | undefined, dflt: number, map: (v: T) => unknown): unknown[] {
	const out: unknown[] = [];
	for (const [k, v] of Object.entries(section ?? {})) {
		const n = keyNums(k);
		if (n.length < 2) continue;
		out.push([bits(n[0]), bits(n[1]), bits(n.length >= 3 ? n[2] : dflt), map(v)]);
	}
	return out;
}

/** `[latBits, lonBits, answer]`. The cascade's `nearbyWays` never passes a
 *  radius (`velocity.ts:1265`), which is why `Env.nearbyWays` takes two
 *  arguments — so only the two-part keys belong here. */
function table2<T>(section: Record<string, T> | undefined, map: (v: T) => unknown): unknown[] {
	const out: unknown[] = [];
	for (const [k, v] of Object.entries(section ?? {})) {
		const n = keyNums(k);
		if (n.length !== 2) continue;
		out.push([bits(n[0]), bits(n[1]), map(v)]);
	}
	return out;
}

/** `reverseGeocode(cityGrid(lat), cityGrid(lon), 16)` — the endpoint city lookup
 *  `enrichMovingSegment` makes, which reaches the fold through
 *  `reenrichSplitWalks`.
 *
 *  `[latBits, lonBits, zoom, address|null]`. The zoom crosses as a PLAIN INTEGER,
 *  unlike every other table key: it is a literal the caller writes (16 here, 18
 *  by default), not a measured double, so keying it on float bits would be a
 *  spelling both sides must agree on for nothing. The coordinates stay on bits
 *  for the usual reason — JS and Lean render doubles differently.
 *
 *  A `null` answer is Nominatim resolving nothing, which is a RESULT. Stored as
 *  such, so a key present with a null answer never reads as a miss.
 *
 *  Only the five address fields `extractCity` reads cross. The rest of a
 *  `NominatimResult` belongs to the venue namers, which the fold does not run. */
function geocodeTable(section: Record<string, NominatimResult | null> | undefined): unknown[] {
	const out: unknown[] = [];
	for (const [k, v] of Object.entries(section ?? {})) {
		const n = keyNums(k);
		if (n.length < 2) continue;
		out.push([
			bits(n[0]),
			bits(n[1]),
			n.length >= 3 ? n[2] : NOMINATIM_DEFAULT_ZOOM,
			v === null
				? null
				: {
						stateDistrict: optStr(v.address.state_district),
						city: optStr(v.address.city),
						town: optStr(v.address.town),
						village: optStr(v.address.village),
						municipality: optStr(v.address.municipality),
					},
		]);
	}
	return out;
}

/** Build the request.
 *
 *  `answers` is what the REPLAY'S OWN adapter answered, not the fixture's
 *  `osmTrace`, and the difference is a defect this used to have. Under
 *  `osmSource: "rows"` — golden's default — the TS arm gets a
 *  `RowSetOsmAdapter` that COMPUTES the four spatial lookups over raw OSM rows,
 *  so its coordinate domain is unbounded. `osmTrace` is a fixed record from an
 *  older capture, and on 2026-06-15 the TS arm asked `nearbyWays` about 69
 *  distinct coordinates of which 4 were not in the trace at all. Building the
 *  Lean tables from the trace therefore handed the fold a strictly SMALLER
 *  oracle and manufactured misses that say nothing about the port (#428).
 *
 *  It is also the second-order fix: the row set and the trace are two different
 *  oracles (#412), so a trace-built table can carry a different ANSWER for a
 *  key both hold, not merely be missing keys. A recorded table cannot.
 *
 *  `trace` asks for per-pass output, which is how a divergence gets attributed
 *  to the pass that produced it (#409) — off by default because it multiplies
 *  the response by 38. */
export function buildDayRequest(cap: FoldCaptureFile, day: CapturedDay, answers: OsmTrace, trace = false): unknown {
	const t = answers;
	const inputs = day.inputs;
	return {
		segs: cap.segsIn.map(encodeSeg),
		trace,
		env: {
			homeTz: inputs.homeTz,
			points: cap.obs.points.map((p) => [p.ts, bits(p.lat), bits(p.lon), bits(p.speedKmh)]),
			rawFixes: cap.obs.rawFixes.map((p) => [p.ts, bits(p.lat), bits(p.lon), optBits(p.accuracy)]),
			displayFixes: cap.obs.displayFixes.map((p) => [p.ts, bits(p.lat), bits(p.lon), optBits(p.accuracy)]),
			steps: cap.obs.steps.map((s) => [s.ts, bits(s.steps)]),
			hr: cap.obs.hr.map((h) => [h.ts, bits(h.bpm)]),
			sleep: cap.obs.sleep.map((s) => [s.startTs, s.endTs]),
			// Kalman speed at a timestamp — a projection of `points`, not a
			// separate observation, so it is derived here rather than captured.
			speedByTs: cap.obs.points.map((p) => [p.ts, bits(p.speedKmh)]),

			knownPlaces: inputs.knownPlaces.map((p) => [p.id, bits(p.centroidLat), bits(p.centroidLon)]),
			focusPlaceDays: inputs.knownPlaces.map((p) => [p.id, p.uniqueDays]),
			hsmmPlaces: inputs.knownPlaces.map((p) => [
				p.id,
				optStr(p.displayName),
				bits(p.centroidLat),
				bits(p.centroidLon),
			]),
			hmmDecode: (inputs.hsmmDecode ?? []).map((h) => ({
				startTs: h.startTs,
				endTs: h.endTs,
				mode: h.mode,
				lineName: optStr(h.lineName),
				placeId: h.placeId ?? null,
			})),
			railRouteCache: inputs.railRouteCache.map((r) => [
				r.routeKey,
				(JSON.parse(r.geometryJson) as { lat: number; lon: number }[]).map((p) => [bits(p.lat), bits(p.lon)]),
			]),
			busRouteCache: (inputs.busRouteCache ?? []).map((b) => ({
				routeRef: b.routeRef,
				routeName: optStr(b.routeName),
				osmRelationId: b.osmRelationId,
				stops: b.stops.map((s) => [optStr(s.name), bits(s.lat), bits(s.lon), s.seq]),
			})),
			railStops: (inputs.railStopsCache ?? []).map((r) => ({
				lineRef: optStr(r.lineRef),
				lineName: optStr(r.lineName),
				osmRelationId: r.osmRelationId,
				routeType: r.routeType,
				stops: r.stops.map((s) => [optStr(s.name), bits(s.lat), bits(s.lon), s.seq]),
			})),

			lookups: {
				nearbyStations: table3(t.nearbyStations, DEFAULT_RADIUS_M.nearbyStations, (v) =>
					v.map((s) => ({
						name: s.name,
						subtype: s.subtype,
						distanceM: bits(s.distanceM),
						lat: optBits(s.lat),
						lon: optBits(s.lon),
					})),
				),
				linesAtPoint: table3(t.linesAtPoint, DEFAULT_RADIUS_M.linesAtPoint, (v) => [...v]),
				transitStops: table3(t.nearbyTransitStops, DEFAULT_RADIUS_M.nearbyTransitStops, (v) =>
					v.map((s) => ({ subtype: s.subtype, distanceM: bits(s.distanceM) })),
				),
				nearbyWays: table2(t.nearbyWays, (v) =>
					v.map((w) => ({
						type: w.type,
						subtype: w.subtype,
						name: optStr(w.name),
						distanceM: optBits(w.distanceM),
					})),
				),
				stationsOnLine: Object.entries(t.stationsOnLine ?? {}).map(([line, v]) => [
					line,
					v.map((s) => [s.name, bits(s.lat), bits(s.lon)]),
				]),
				reverseGeocode: geocodeTable(t.reverseGeocode),
				// The two the adapter never saw — see `fold-capture.ts`.
				tzAt: cap.tzAt.map((q) => [bits(q.lat), bits(q.lon), q.tz]),
				bestPlace: cap.bestPlace.map((q) => [
					bits(q.lat),
					bits(q.lon),
					q.startTs,
					q.endTs,
					q.tz,
					q.label === null ? null : { label: q.label, city: q.city },
				]),
			},
		},
	};
}

File: rust/backend/src/fold_payload.rs
//! The day fold's wire encoding — port of `src/lean/fold-payload.ts` (#982).
//!
//! The fold takes one JSON object and the pipeline builds it. Under
//! `LEAN_DAY=solo` — which the decode CronJob runs — this encoding IS the
//! interface between the TypeScript pipeline and the verified core, so a Rust
//! host that gets it wrong asks a different question and gets a confidently
//! wrong day rather than an error.
//!
//! # Floats cross as their BIT PATTERN, and that is the point
//!
//! `bits` renders an `f64` as the decimal of its IEEE-754 payload, so the
//! double the fold reads is the double the pipeline had — not a rendering of
//! it. Anything that formats the number instead (`{}`, `{:.6}`, `to_string` on
//! the value) is a silent precision change on coordinates, where 1e-7° is the
//! quantisation the match gate already adjudicates.
//!
//! # Verified against the TypeScript's own output, not against my reading
//!
//! `tests/fold_payload.rs` compares this to `dist/lean/fold-payload.js`'s
//! encoding of the same captured day, byte for byte. That matters more here
//! than in most ports: the shape is 25 keys of mostly-pass-through data where a
//! wrong field name produces a well-formed request the fold answers anyway.

use anyhow::{Context, Result};
use serde_json::{Map, Value, json};

/// An `f64` as the decimal of its IEEE-754 bit pattern.
///
/// The inverse of `day-serve.ts`'s `unbits`. ⚠ NaN has many bit patterns and
/// this preserves whichever one it was handed, exactly as the TypeScript's
/// `DataView` round trip does — it is a transport, not a normaliser.
pub fn bits(x: f64) -> String {
    x.to_bits().to_string()
}

/// `bits`, but `null` survives as `null`.
///
/// ⚠ Absent and zero are different facts here. A segment with no `refinedMode`
/// is not a segment refined to nothing, and encoding one as the other is the
/// class of defect the activity-summary port already paid for.
pub fn opt_bits(x: Option<f64>) -> Value {
    match x {
        Some(v) => Value::String(bits(v)),
        None => Value::Null,
    }
}

/// A JSON number field as `bits`, or `null` when absent or null.
fn num_bits(o: &Map<String, Value>, k: &str) -> Value {
    opt_bits(o.get(k).and_then(Value::as_f64))
}

/// A JSON field passed through as a string, or `null` when absent or null.
fn opt_str(o: &Map<String, Value>, k: &str) -> Value {
    match o.get(k) {
        Some(Value::String(s)) => Value::String(s.clone()),
        _ => Value::Null,
    }
}

/// A field passed through UNENCODED.
///
/// Timestamps, counts and ids are not bit-encoded: JSON carries an integer
/// exactly, so there is nothing to preserve. It also carries the string keys —
/// `routeKey`, and the enum-ish `mode` values — which is why this is `raw`
/// rather than `int`: the TypeScript writes `r.routeKey` with no conversion at
/// all, and naming it for one of the types it passes would invite someone to
/// "fix" the others.
fn raw(o: &Map<String, Value>, k: &str) -> Value {
    o.get(k).cloned().unwrap_or(Value::Null)
}

/// A `{lat, lon, ts}` path as the fold's triples of bit patterns, or `null`.
fn path(v: Option<&Value>) -> Value {
    match v.and_then(Value::as_array) {
        None => Value::Null,
        Some(pts) => Value::Array(
            pts.iter()
                .map(|p| {
                    let o = p.as_object();
                    let g = |k: &str| {
                        o.and_then(|o| o.get(k))
                            .and_then(Value::as_f64)
                            .map(bits)
                            .map_or(Value::Null, Value::String)
                    };
                    json!([g("lat"), g("lon"), g("ts")])
                })
                .collect(),
        ),
    }
}

/// One segment in the shape `Day.parseSeg` reads.
///
/// ⚠ THIRTY FIELDS, and the tail of the list is the easy half to miss — my
/// first attempt stopped at `wayName` because that is where the first screen
/// of the TypeScript ends. A short encoding is not a parse error on the Lean
/// side: absent fields take `Seg`'s defaults, so the fold answers a
/// well-formed question about a segment that has lost its enrichment.
pub fn encode_seg(seg: &Value) -> Value {
    let Some(o) = seg.as_object() else {
        return Value::Null;
    };
    json!({
        "startTs": raw(o, "startTs"),
        "endTs": raw(o, "endTs"),
        "mode": opt_str(o, "mode"),
        "refinedMode": opt_str(o, "refinedMode"),
        "confidence": num_bits(o, "confidence"),
        "confidenceMargin": num_bits(o, "confidenceMargin"),
        "avgSpeed": num_bits(o, "avgSpeed"),
        "maxSpeed": num_bits(o, "maxSpeed"),
        "linearity": num_bits(o, "linearity"),
        "pointCount": raw(o, "pointCount"),
        "place": opt_str(o, "place"),
        "city": opt_str(o, "city"),
        "wayName": opt_str(o, "wayName"),
        "refinedReason": opt_str(o, "refinedReason"),
        // Absent becomes `[]`, not `null`: Lean reads a list and the TypeScript
        // spreads or substitutes an empty array.
        "refinedKinds": o.get("refinedKinds").cloned().unwrap_or_else(|| json!([])),
        "centroidLat": num_bits(o, "centroidLat"),
        "centroidLon": num_bits(o, "centroidLon"),
        // ⚠ `string | number` in TypeScript, `Option Int` in Lean. A
        // non-numeric id would become NaN, so it is DROPPED rather than
        // coerced — the TypeScript's own choice, kept.
        "focusPlaceId": focus_place_id(o.get("focusPlaceId")),
        "needsReenrich": o.get("needsReenrich").and_then(Value::as_bool).unwrap_or(false),
        "needsRename": o.get("needsRename").and_then(Value::as_bool).unwrap_or(false),
        "vehicleKind": opt_str(o, "vehicleKind"),
        "roadCorridorFraction": num_bits(o, "roadCorridorFraction"),
        "displayTz": opt_str(o, "displayTz"),
        "snappedPath": path(o.get("snappedPath")),
        "matchedPath": path(o.get("matchedPath")),
        "walkMatchedPath": path(o.get("walkMatchedPath")),
        "walkSmoothedPath": path(o.get("walkSmoothedPath")),
        "biometrics": encode_biometrics(o.get("biometrics")),
    })
}

/// `string | number | undefined` → `Option Int`. Anything not a number is
/// dropped rather than coerced to NaN.
fn focus_place_id(v: Option<&Value>) -> Value {
    match v {
        Some(Value::Number(n)) => Value::Number(n.clone()),
        Some(Value::String(s)) => s
            .parse::<f64>()
            .ok()
            .filter(|f| !f.is_nan())
            .and_then(serde_json::Number::from_f64)
            .map_or(Value::Null, Value::Number),
        _ => Value::Null,
    }
}

/// The per-segment biometric summary, or `null` when the segment carries none.
///
/// ⚠ `sleepFraction` is `bits`, NOT `optBits` — the TypeScript reads it
/// unconditionally, so a segment with biometrics always has one.
fn encode_biometrics(v: Option<&Value>) -> Value {
    let Some(b) = v.and_then(Value::as_object) else {
        return Value::Null;
    };
    json!({
        "hrMean": num_bits(b, "hrMean"),
        "hrMin": num_bits(b, "hrMin"),
        "hrMax": num_bits(b, "hrMax"),
        "hrStd": num_bits(b, "hrStd"),
        "sampleCount": raw(b, "sampleCount"),
        "overlapsSleep": b.get("overlapsSleep").cloned().unwrap_or(Value::Null),
        "sleepFraction": num_bits(b, "sleepFraction"),
        "stepsTotal": num_bits(b, "stepsTotal"),
    })
}

/// A `[ts, bits(lat), bits(lon)]` fixture list — the shape both cross-day
/// fixture arrays take.
fn fixes3(v: Option<&Value>) -> Value {
    arr_map(v, |o| {
        json!([raw(o, "ts"), num_bits(o, "lat"), num_bits(o, "lon")])
    })
}

/// A `[ts, bits(lat), bits(lon), optBits(accuracy)]` fixture list.
fn fixes4(v: Option<&Value>) -> Value {
    arr_map(v, |o| {
        json!([
            raw(o, "ts"),
            num_bits(o, "lat"),
            num_bits(o, "lon"),
            num_bits(o, "accuracy")
        ])
    })
}

/// Map over a JSON array of objects, skipping anything that is not one.
///
/// An absent array becomes `[]`, not `null`: every one of these fields is a
/// list on the Lean side, and the TypeScript spells the same default with
/// `?? []`.
fn arr_map<F: Fn(&Map<String, Value>) -> Value>(v: Option<&Value>, f: F) -> Value {
    Value::Array(
        v.and_then(Value::as_array)
            .map(|a| a.iter().filter_map(Value::as_object).map(&f).collect())
            .unwrap_or_default(),
    )
}

/// The observation series and the cross-day tail, as `env` carries them.
///
/// ⚠ `points` and `speedByTs` are the SAME source read twice, and both are
/// sent: the fold wants the track and the speed lookup separately, and
/// deriving one from the other here would be this encoder deciding something.
///
/// ⚠ `sleep` and `rawSleep` are NOT the same list. `sleep` is the projection
/// the biometric windows read; `rawSleep` is the Fitbit windows before place
/// attribution — same rows, different fields, and neither derives the other.
pub fn encode_obs_and_tail(obs: &Value, tail: Option<&Value>) -> Map<String, Value> {
    let o = obs.as_object();
    let g = |k: &str| o.and_then(|o| o.get(k));
    let t = tail.and_then(Value::as_object);
    let tg = |k: &str| t.and_then(|t| t.get(k));

    let mut m = Map::new();
    m.insert(
        "points".into(),
        arr_map(g("points"), |p| {
            json!([
                raw(p, "ts"),
                num_bits(p, "lat"),
                num_bits(p, "lon"),
                num_bits(p, "speedKmh")
            ])
        }),
    );
    m.insert("rawFixes".into(), fixes4(g("rawFixes")));
    m.insert("displayFixes".into(), fixes4(g("displayFixes")));
    m.insert(
        "steps".into(),
        arr_map(g("steps"), |s| json!([raw(s, "ts"), num_bits(s, "steps")])),
    );
    m.insert(
        "hr".into(),
        arr_map(g("hr"), |h| json!([raw(h, "ts"), num_bits(h, "bpm")])),
    );
    m.insert(
        "sleep".into(),
        arr_map(g("sleep"), |s| json!([raw(s, "startTs"), raw(s, "endTs")])),
    );
    m.insert(
        "speedByTs".into(),
        arr_map(g("points"), |p| {
            json!([raw(p, "ts"), num_bits(p, "speedKmh")])
        }),
    );
    m.insert("morningFixes".into(), fixes3(tg("morningRaw")));
    m.insert("prevEveningFixes".into(), fixes3(tg("prevEveningRaw")));
    m.insert(
        "rawSleep".into(),
        arr_map(tg("rawSleep"), |w| {
            json!([
                raw(w, "startTs"),
                raw(w, "endTs"),
                opt_str(w, "tz"),
                raw(w, "minutesAsleep")
            ])
        }),
    );
    m.insert(
        "dayEndTs".into(),
        tg("dayEndTs").cloned().unwrap_or_else(|| json!(0)),
    );
    m
}

/// The mined `mode_biometrics` rows, as the eleven-tuple `env.modeStats` takes.
pub fn encode_mode_stats(v: Option<&Value>) -> Value {
    arr_map(v, |m| {
        json!([
            opt_str(m, "mode"),
            num_bits(m, "hrMean"),
            num_bits(m, "hrStd"),
            raw(m, "hrSampleCount"),
            num_bits(m, "cadenceMean"),
            num_bits(m, "cadenceStd"),
            raw(m, "cadenceSampleCount"),
            num_bits(m, "speedMean"),
            num_bits(m, "speedStd"),
            raw(m, "speedSampleCount"),
            raw(m, "sampleCount"),
        ])
    })
}

/// A required float field: `bits`, with a default when absent.
///
/// ⚠ Distinct from [`num_bits`] and the difference is load-bearing. These
/// fields are `bits(x ?? d)` in the TypeScript — a number Lean always reads —
/// whereas `num_bits` is `optBits`, where absent stays `null`. Using the wrong
/// one turns "unknown" into a specific value or vice versa.
fn num_bits_or(o: &Map<String, Value>, k: &str, dflt: f64) -> Value {
    Value::String(bits(o.get(k).and_then(Value::as_f64).unwrap_or(dflt)))
}

/// The mined places, SIX WAYS, as `env` carries them.
///
/// ⚠ One source list, six different tuples, and they are not redundant: each
/// stage reads the columns it needs and Lean indexes them positionally. The
/// stay DETECTOR wants a display name to snap a cluster to; the dwell
/// CONTINUATION wants visit counts; the enrichment stage wants the whole row
/// because it scores geometry and hour profile and then branches the LABEL on
/// the same row. Emitting five of the six, or the right six in the wrong
/// order, is not a decode error on the far side.
pub fn encode_places(known: Option<&Value>) -> Map<String, Value> {
    let mut m = Map::new();
    m.insert(
        "stayPlaces".into(),
        arr_map(known, |p| {
            json!([
                num_bits(p, "centroidLat"),
                num_bits(p, "centroidLon"),
                num_bits(p, "radiusM"),
                opt_str(p, "displayName")
            ])
        }),
    );
    m.insert(
        "dwellPlaces".into(),
        arr_map(known, |p| {
            json!([
                num_bits(p, "centroidLat"),
                num_bits(p, "centroidLon"),
                num_bits(p, "radiusM"),
                num_bits(p, "totalDwellSec"),
                // ⚠ `?? null` in the TypeScript, NOT `?? 0`: a place with no
                // recorded visits is not a place visited zero times.
                p.get("visitCount").cloned().unwrap_or(Value::Null),
                raw(p, "uniqueDays")
            ])
        }),
    );
    m.insert(
        "enrichPlaces".into(),
        arr_map(known, |p| {
            json!([
                raw(p, "id"),
                num_bits(p, "centroidLat"),
                num_bits(p, "centroidLon"),
                // ⚠ Defaults to 50 m here and stays null in `stayPlaces` above.
                // Both are the TypeScript's, on the same field.
                num_bits_or(p, "radiusM", 50.0),
                num_bits_or(p, "uniqueDays", 0.0),
                hour_profile(p.get("hourProfile")),
                opt_str(p, "displayName"),
                num_bits_or(p, "sleepHours", 0.0),
                opt_str(p, "amenityLabel")
            ])
        }),
    );
    m.insert(
        "knownPlaces".into(),
        arr_map(known, |p| {
            json!([
                raw(p, "id"),
                num_bits(p, "centroidLat"),
                num_bits(p, "centroidLon")
            ])
        }),
    );
    m.insert(
        "focusPlaceDays".into(),
        arr_map(known, |p| json!([raw(p, "id"), raw(p, "uniqueDays")])),
    );
    m.insert(
        "hsmmPlaces".into(),
        arr_map(known, |p| {
            json!([
                raw(p, "id"),
                opt_str(p, "displayName"),
                num_bits(p, "centroidLat"),
                num_bits(p, "centroidLon")
            ])
        }),
    );
    m
}

/// The 24-hour visit profile: every entry `bits`, or `null` for the whole
/// thing. ⚠ `null` means "no profile", which is not a profile of zeroes.
fn hour_profile(v: Option<&Value>) -> Value {
    match v.and_then(Value::as_array) {
        None => Value::Null,
        Some(hs) => Value::Array(hs.iter().map(|h| opt_bits(h.as_f64()).clone()).collect()),
    }
}

/// A route's stop list: `[optStr(name), bits(lat), bits(lon), seq]`.
///
/// Shared by the bus and rail caches — the same four columns in the same order,
/// which is why one helper serves both rather than two that could drift.
fn stops(v: Option<&Value>) -> Value {
    arr_map(v, |s| {
        json!([
            opt_str(s, "name"),
            num_bits(s, "lat"),
            num_bits(s, "lon"),
            raw(s, "seq")
        ])
    })
}

/// The route and decode caches, as `env` carries them.
///
/// ⚠ `railRouteCache` holds its geometry as a JSON STRING inside the row, so
/// this parses a document out of a field, and a row that does not parse is an
/// ERROR rather than a row that is quietly skipped.
///
/// The first version dropped it and called that a deliberate improvement on the
/// TypeScript, which throws. `dev-lint`'s `rust-serde-swallow` rejected it, and
/// the linter was right: a corrupt cache row silently becoming "this route does
/// not exist" is a matcher scoring a day against a route set that is missing
/// something, with nothing anywhere saying so. The three outcomes are not
/// symmetric — an empty polyline is a route that goes nowhere, a dropped row is
/// a route nobody offered, and only an error is a statement about the cache.
pub fn encode_caches(
    hsmm_decode: Option<&Value>,
    rail_routes: Option<&Value>,
    bus_routes: Option<&Value>,
    rail_stops: Option<&Value>,
) -> Result<Map<String, Value>> {
    let mut m = Map::new();

    m.insert(
        "hmmDecode".into(),
        arr_map(hsmm_decode, |h| {
            json!({
                "startTs": raw(h, "startTs"),
                "endTs": raw(h, "endTs"),
                "mode": opt_str(h, "mode"),
                "lineName": opt_str(h, "lineName"),
                // `?? null`: an absent place id is unknown, not place zero.
                "placeId": h.get("placeId").cloned().unwrap_or(Value::Null),
            })
        }),
    );

    let mut routes = Vec::new();
    for row in rail_routes.and_then(Value::as_array).into_iter().flatten() {
        let r = row
            .as_object()
            .context("railRouteCache row is not an object")?;
        let geom_str = r
            .get("geometryJson")
            .and_then(Value::as_str)
            .context("railRouteCache row has no geometryJson string")?;
        let geom: Value = serde_json::from_str(geom_str).with_context(|| {
            format!(
                "railRouteCache row {} has unparseable geometryJson",
                r.get("routeKey").and_then(Value::as_str).unwrap_or("?")
            )
        })?;
        routes.push(json!([
            raw(r, "routeKey"),
            arr_map(Some(&geom), |p| json!([
                num_bits(p, "lat"),
                num_bits(p, "lon")
            ]))
        ]));
    }
    m.insert("railRouteCache".into(), Value::Array(routes));

    m.insert(
        "busRouteCache".into(),
        arr_map(bus_routes, |b| {
            json!({
                "routeRef": raw(b, "routeRef"),
                "routeName": opt_str(b, "routeName"),
                "osmRelationId": raw(b, "osmRelationId"),
                "stops": stops(b.get("stops")),
            })
        }),
    );

    m.insert(
        "railStops".into(),
        arr_map(rail_stops, |r| {
            json!({
                "lineRef": opt_str(r, "lineRef"),
                "lineName": opt_str(r, "lineName"),
                "osmRelationId": raw(r, "osmRelationId"),
                "routeType": raw(r, "routeType"),
                "stops": stops(r.get("stops")),
            })
        }),
    );

    Ok(m)
}
